c++/src/capnp/compat/json.c++

```diff
@@ -380,72 +380,135 @@ void JsonCodec::encodeField(StructSchema::Field field, DynamicValue::Reader inpu
 }
 
 namespace {
-int64_t parseInt64(kj::StringPtr s)
-{
+int64_t parseInt64(kj::StringPtr s) {
   char *endPtr;
   errno = 0;
   int64_t value = std::strtoll(s.begin(), &endPtr, 10);
-  KJ_REQUIRE(endPtr == s.end() && errno != ERANGE, "String is not correct int64 number");
+  KJ_REQUIRE(endPtr == s.end(), "String does not contain valid number", s);
+  KJ_REQUIRE(errno != ERANGE, "Value out-of-range", s);
   return value;
 }
 
-uint64_t parseUInt64(kj::StringPtr s)
-{
+uint64_t parseUInt64(kj::StringPtr s) {
   char *endPtr;
   errno = 0;
   uint64_t value = std::strtoull(s.begin(), &endPtr, 10);
-  KJ_ASSERT(endPtr == s.end() && errno != ERANGE, "String is not correct uint64 number");
+  KJ_REQUIRE(endPtr == s.end(), "String does not contain valid number", s);
+  KJ_REQUIRE(errno != ERANGE, "Value out-of-range", s);
   return value;
 }
 
-template <typename Set, typename DecodeArray, typename DecodeObject>
-void decodeField(Type type, JsonValue::Reader value, Set set, DecodeArray decodeArray,
-    DecodeObject decodeObject) {
-  switch (value.which()) {
-    case JsonValue::NULL_:
+double parseFloat64(kj::StringPtr s) {
+  char *endPtr;
+  errno = 0;
+  double value = std::strtod(s.begin(), &endPtr);
+  KJ_REQUIRE(endPtr == s.end(), "String does not contain valid floating number", s);
+  return value;
+}
+
+template <typename SetFn, typename DecodeArrayFn, typename DecodeObjectFn>
+void decodeField(Type type, JsonValue::Reader value, SetFn setFn, DecodeArrayFn decodeArrayFn,
+    DecodeObjectFn decodeObjectFn) {
+  //This code relies on conversions in DynamicValue::Reader::as<T>.
+  switch(type.which()) {
+    case schema::Type::VOID:
+      break;
+    case schema::Type::BOOL:
+      if (value.isBoolean()) {
+        setFn(value.getBoolean());
+      } else {
+        KJ_FAIL_REQUIRE("Expected boolean value");
+      }
       break;
-    case JsonValue::BOOLEAN:
-      set(value.getBoolean());
+    case schema::Type::INT8:
+    case schema::Type::INT16:
+    case schema::Type::INT32:
+    case schema::Type::INT64:
+      //Rellies on range check in DynamicValue::Reader::as<IntType>
+      if (value.isNumber()) {
```

Comment:
rather than a sequence of `if / else if / ... / else` here, you could `switch(value.which())`. It *might* produce more performant code, but I think it will certainly read better.